### PR TITLE
chore: Use Ops.CollectStatus Event

### PR DESCRIPTION
**Pull request by gatici**
# Description

This PR uses new ops feature CollectStatus Event to manage status Charm. Event named collect_unit_status sets the status of charm automatically at the end of every hook.

Reference:

    https://ops.readthedocs.io/en/latest/#ops.CollectStatusEvent
    https://discourse.charmhub.io/t/how-to-set-a-charms-status/11771

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -84,37 +85,87 @@ def __init__(self, *args):
...
+
+        """
+        if not self._container.can_connect():
+            return WaitingStatus("Waiting for the container to be ready")
+        if not self._storage_is_attached():
+            return WaitingStatus("Waiting for the storage to be attached")
+        if not _get_pod_ip():
+            return WaitingStatus("Waiting for pod IP address to be available")
+
+        return None
```
Comment (gruyaume) [inline src/charm.py:137]: I'm not a fan of those kinds of methods. The name implies that we would return a bool, not a status.

File: src/charm.py [hunk 2]
```diff
@@ -84,37 +85,87 @@ def __init__(self, *args):
...
+        """
+        if status := self._is_unit_in_non_active_status():
+            event.add_status(status)
+            return
+
+        if self._csr_is_stored() and not self._get_current_provider_certificate():
+            event.add_status(WaitingStatus("Waiting for certificates to be stored"))
+            return
+
+        event.add_status(ActiveStatus())
```
Comment (gruyaume) [inline src/charm.py:168]: It feels like we are making things more complex by adding all those new methods. Can't we simply have everything status related in this `_on_collect_unit_status` method?

```python
if not xyz:
    event.add_status(BlockedStatus("xyz")
if not abc:
    ...
event.add_status(ActiveStatus())
```

Example in Vault K8s: 
- https://github.com/canonical/vault-k8s-operator/blob/main/src/charm.py#L164

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 1 above)
Comment (gatici) [reply src/charm.py:137]: I agree, the method name could be changed. However if we put everything into  _on_collect_unit_status method, we get  C901 'UDROperatorCharm._on_collect_unit_status' is too complex linting error.  If disabling C901 is fine, I can do that.

File: src/charm.py (hunk 2 above)
Comment (gatici) [reply src/charm.py]: Done

File: src/charm.py (hunk 1 above)
Comment (gruyaume) [reply src/charm.py]: I know it makes for a longer method but it's way more readable and should make our life much simpler when problems occur here.

**Review by gruyaume — COMMENTED** [review]
Looks much better, only 1 more comment.

File: src/charm.py [hunk 3]
```diff
@@ -84,35 +85,87 @@ def __init__(self, *args):
...
+            return
+
+        if not _get_pod_ip():
+            event.add_status(WaitingStatus("Waiting for pod IP address to be available"))
+            return
+
+        if self._csr_is_stored() and not self._get_current_provider_certificate():
+            event.add_status(WaitingStatus("Waiting for certificates to be stored"))
+            return
+
```
[the remark this answers is not in the document]
Comment (gatici) [reply src/charm.py:150]: Added.

[the remark this answers is not in the document]
Comment (gatici) [reply src/charm.py:150]: Done

**Review by gruyaume — APPROVED** [review]

**Pull request merged**